ice-adapter/src/main/java/com/faforever/iceadapter/ice/PeerIceModule.java

```diff
@@ -153,11 +158,50 @@ private void gatherCandidates() {
     //How often have we been waiting for a response to local candidates/offer
     private volatile int awaitingCandidatesEventId = 0;
 
+    private List<IceServer> getViableIceServers() {
+        List<IceServer> allIceServers = GameSession.getIceServers();
+        if (IceAdapter.PING_COUNT <= 0 || allIceServers.isEmpty()) {
+            return allIceServers;
+        }
+
+        List<IceServer> fafIceServers = allIceServers.stream()
+                .filter(server -> server.getTurnAddresses().stream()
+                        .anyMatch(transportAddress -> transportAddress.getHostString().contains("faforever.com")))
+                .collect(Collectors.toList());
+        // Try official servers first
+        List<IceServer> viableIceServers = fafIceServers.stream()
+                .filter(IceServer::hasAcceptableLatency)
+                .collect(Collectors.toList());
+        if (!viableIceServers.isEmpty()) {
+            return viableIceServers;
+        }
+
+        // Try servers with acceptable latency
+        viableIceServers = allIceServers.stream()
+                .filter(IceServer::hasAcceptableLatency)
+                .collect(Collectors.toList());
+        if (!viableIceServers.isEmpty()) {
+            return viableIceServers;
+        }
+
+        // Try the closest server
+        IceServer closestIceServer = allIceServers.stream()
+                .min(Comparator.comparing(server -> server.getRoundTripTime().join().orElse(Double.POSITIVE_INFINITY))).get();
+        if (closestIceServer.getRoundTripTime().join().isPresent()) {
+            viableIceServers.add(closestIceServer);
+        }
```

Comment:
I'd go for \/ to avoid the `Double.POSITIVE_INFINITY`.
```java
Optional<IceServer> closestIceServer = allIceServers.stream()
    .filter(server.getRoundTripTime().get().isPresent())
    .min(Comparator.comparing(server -> server.getRoundTripTime();
if (closestIceServer.isPresent()) {
    ...
}

```